src/wp-includes/html-api/class-wp-html-processor.php

```diff
@@ -406,26 +406,66 @@ public function next_tag( $query = null ) {
 		$breadcrumbs  = $query['breadcrumbs'];
 		$match_offset = isset( $query['match_offset'] ) ? (int) $query['match_offset'] : 1;
 
-		$crumb  = end( $breadcrumbs );
-		$target = strtoupper( $crumb );
 		while ( $match_offset > 0 && $this->step() ) {
-			if ( $target !== $this->get_tag() ) {
-				continue;
+			if ( $this->matches_breadcrumbs( $breadcrumbs ) && 0 === --$match_offset ) {
+				return true;
 			}
+		}
 
-			// Look up the stack to see if the breadcrumbs match.
-			foreach ( $this->state->stack_of_open_elements->walk_up() as $node ) {
-				if ( strtoupper( $crumb ) !== $node->node_name ) {
-					break;
-				}
+		return false;
+	}
 
-				$crumb = prev( $breadcrumbs );
-				if ( false === $crumb && 0 === --$match_offset && ! $this->is_tag_closer() ) {
-					return true;
-				}
+	/**
+	 * Indicates if the currently-matched tag matches the given breadcrumbs.
+	 *
+	 * A "*" represents a single tag wildcard, where any tag matches, but not no tags.
+	 *
+	 * At some point this function _may_ support a `**` syntax for matching any number
+	 * of unspecified tags in the breadcrumb stack. This has been intentionally left
+	 * out, however, to keep this function simple and to avoid introducing backtracking,
+	 * which could open up surprising performance breakdowns.
```

Comment:
ha yeah this would work and accurately conveys the backtracking problem, though I believe that there are some more performant backtracking mechanisms that don't have to re-process as much as this.

Reply:
it's still possible that I'm way overthinking this, confusing it with a different kind of match, where it would matter which of the inner or outer elements are matching.

for a pattern like `main, **, div, **, img` though I want to make sure this doesn't break, or maybe it's better to compare that against `main, *, div, **, img`

for the first one, with two `**` groups, I read it as "walk up the stack of open elements; if there is a DIV somewhere and also a MAIN somewhere then this matches."

for the second one though we have a tighter constraint above the simpler one. if we have this breadcrumb `HTML > BODY > MAIN > SECTION > DIV > DIV > IMG` then we have to make sure that the `**` group _only_ matches the last DIV, otherwise `main, *, div` will fail to match because we were too greedy.

This is where I think we must start backtracking: after that higher DIV fails the rest of the match. That's a very recursive problem too. If we start by being un-greedy though, we stop at the lower DIV and it fails the rest of the match, so we have to track that `**` group and backtrack on the pattern vs. on the content. We have to remember each double wildcard group and go back to reattempt matches after adjusting our position in the stack.

The probably probably starts if `*` allowed for a zero match, but when I tossed it in there it didn't seem to have any connotation of that since this is like a filesystem wildcard where those expectations are that `*` in a path is _something_ and not _nothing_.

Reply:
thinking all this through pushes me to leave this for userspace for now. we expose `get_breadcrumbs()` and so someone is implement this on their own until we find a good solution for it.

```php
function preg_matches_breadcrumbs( string $pattern, array $breadcrumbs ) {
	return false !== preg_match( $pattern, implode( ' > ', $breadcrumbs ) );
}

if ( preg_matches_breadcrumbs( '~MAIN > .* IMG~', $processor->get_breadcrumbs() ) ) {
	…
}
```